### ISSUE TITLE
Improve handling of clock skew

### ISSUE THREAD

**Issue by mattheworiordan** [ably/ably-dotnet#303]
See `RSA4b1` as part of 1.2 spec changes in https://github.com/ably/docs/pull/622/files

Given this impacts users now with significant clock skew, we aim to release a fix for this now.

**Comment by sacOO7** [comment]
Is this handled in another client's implementation? If it is, it will be great if I can get a reference to it @QuintinWillison

**Comment by sacOO7** [comment]
Also I cant access intercom conversation regarding issue https://github.com/ably/docs/issues/619

**Comment by sacOO7** [comment]
Can we get this conversation link @QuintinWillison , Also it will be great if you can help me with reproducing the issue.

**Comment by QuintinWillison** [comment]
@sacOO7 it is implemented elsewhere, yes:

- https://github.com/ably/ably-js/pull/583
- https://github.com/ably/ably-java/pull/463

In terms of "reproducing the issue" see the tests added to the JavaScript and Java implementations.

I am, also, unable to access that Intercom conversation link. I'm not sure what's going on there. I've reached out to our internal tools help team for assistance ([internal request](https://ably-real-time.slack.com/archives/C010R6HD2U8/p1597825937007600)).

**Comment by sacOO7** [comment]
Did we get access to intercom conversation?

**Comment by QuintinWillison** [comment]
No, I'm afraid not, so this is what we've got for now.

**Comment by SimonWoolf** [comment]
@QuintinWillison another customer is blocked by this issue https://support.ably.com/a/tickets/114335 .

NB: the key word in RSA4b1 is "optionally". @sacOO7 that means that there is a very simple fix that satisfies the spec, which is to just get rid of all `Token.IsExpired` checks. That means that, assuming the library implements RSA4b correctly, it can just wait for ably to tell it if a token is expired, and discard and get a new one then.

**Comment by sacOO7** [comment]
Ok, I will look into it : )

**Comment by sacOO7** [comment]
Hi @SimonWoolf , arent we supposed to maintain server offset wrt client locally and then check if the token expired?

**Comment by sacOO7** [comment]
@marto83 Can you check if the issue is related to https://github.com/ably/ably-dotnet/pull/376 (possible fix already implemented or we need tweaking change for this)

**Comment by sacOO7** [comment]
From this comment, https://github.com/ably/ably-dotnet/pull/376#pullrequestreview-297465250 
It seems we deliberately added token expiry wrt local time without offset, it might be causing the issue.

**Comment by SimonWoolf** [comment]
> arent we supposed to maintain server offset wrt client local and then check if the token expired, before making a request?  I am assuming, that's what we have in other implementations.

@sacOO7 the source of truth is the spec; [RSA4b1 says](https://docs.ably.io/client-lib-development-guide/features/#RSA4b1):

`Client libraries can optionally save a round-trip request to the Ably service for expired tokens by detecting when a token has expired when all of the following applies: the current token is a TokenDetails object with an expires attribute; the library has previously queried the time from the Ably service and persisted the local clock offset according to RSA10k; the expires time has passed based on the Ably service time and not the local clock (which is not guaranteed to be accurate)`

Hence my previous comment: the "optionally" there means that you can become spec-compliant just by not checking token expiry. You don't have to do what other implementations do, it's optional. But if you do do it, you must fulfil the rest of the requirements listed there; the problem comes at the moment because it's not doing that. (Not the library's fault, the spec was changed, precisely to fix this issue).

That means that, assuming the library implements RSA4b correctly, it can just wait for ably to tell it if a token is expired, and discard and get a new one then.

I don't believe #376 added token expiry where it wasn't before. It just used the time with offset _if an offset exists_, so it wasn't a regression, which is why I approved it with the comment "happy for it to be merged as-is and that to be done in a separate PR".

**Comment by SimonWoolf** [comment]
(While doing this, we should check that if realtime does return 4014x, that the library does then discard its stored token and fetch a new one. It really should, that isn't something that changed recently in the spec. But worth double-checking because at a quick glance of AblyAuth#AuthorizeAsync it's not obvious to me that it's actually doing that)

**Comment by sacOO7** [comment]
@SimonWoolf Thanks for the info, it's helpful

**Comment by marto83** [comment]
I can confirm we handle the token error codes from realtime both when trying to establish a connection and when we get a Disconnected message. It should be fine to remove the check.

**Comment by sacOO7** [comment]
@marto83 I am assuming we are moving away from (RSA4b1) for now

**Comment by marto83** [comment]
I don't think we should. Querying server time is already there. It just needs a few more tests and tweaks to match with the spec item. 
I'd only disable the token check if you need to do a quick release to fix a customer's issue. I'm going to be working on Ably next week so I think we should address it.

**Comment by sacOO7** [comment]
+1 . Yes, that's what I think. Few tweaks should be able to fix it. ( Btw,
you can go through above conversation by @simon). I feel it's more efforts
to remove the spec and then refactor the code.
